contracts/VestingEscrow.vy

```diff
@@ -0,0 +1,340 @@
+        return 0
+    return self.total_locked - self._total_vested_at(time)
+
+
+@external
+@view
+def locked() -> uint256:
+    """
+    @notice Get the number of locked tokens for recipient
+    """
+    return self._locked(min(block.timestamp, self.disabled_at))
+
+
+@external
+def claim(
+    beneficiary: address = msg.sender, amount: uint256 = max_value(uint256)
+):
+    """
+    @notice Claim tokens which have vested
+    @param beneficiary Address to transfer claimed tokens to
+    @param amount Amount of tokens to claim
+    """
+    self._check_sender_is_recipient()
+
+    claim_period_end: uint256 = min(block.timestamp, self.disabled_at)
+    claimable: uint256 = min(self._unclaimed(claim_period_end), amount)
+    self.total_claimed += claimable
+
+    assert self.token.transfer(beneficiary, claimable)
+    log Claim(self.recipient, beneficiary, claimable)
+
+
+@external
+def revoke_unvested():
+    """
+    @notice Disable further flow of tokens and revoke the unvested part to owner
+    """
+    self._check_sender_is_owner_or_manager()
+    # NOTE: Revoking more than once is futile
+
+    revokable: uint256 = self._locked()
+    self.disabled_at = block.timestamp
+
+    assert self.token.transfer(self.owner, revokable)
+    log RevokeUnvested(self.recipient, revokable)
+
+
+@external
+def change_owner(owner: address):
+    """
+    @notice Change contract owner.
+    @param owner Address of the new owner. Must be non-zero.
+    """
+    self._check_sender_is_owner()
+    assert owner != empty(address), "zero owner address"
+
+    self.owner = owner
+    log OwnerChanged(owner)
+
+
+@external
+def change_manager(manager: address):
+    """
+    @notice Set contract manager.
+            Can update manager if it is already set.
+            Can be called only by the owner.
+    @param manager Address of the new manager
+    """
+    self._check_sender_is_owner()
+
+    self.manager = manager
+    log ManagerChanged(manager)
+
+
+@external
+def revoke_ownership():
+    """
+    @notice Revoke contract owner and manager
+    """
+    self._check_sender_is_owner()
+
+    self.manager = empty(address)
+    self.owner = empty(address)
+
+    log ManagerChanged(empty(address))
+    log OwnerChanged(empty(address))
+
+
+@external
+def recover_erc20(token: address, amount: uint256):
+    """
+    @notice Recover ERC20 tokens to recipient
+    @param token Address of the ERC20 token to be recovered
+    """
+    recoverable: uint256 = amount
+    if token == self.token.address:
+        available: uint256 = ERC20(token).balanceOf(self) - (
+            self._locked() + self._unclaimed()
+        )
+        recoverable = min(recoverable, available)
+    assert ERC20(token).transfer(self.recipient, recoverable)
+    log ERC20Recovered(token, recoverable)
+
+
+@external
+def recover_ether():
+    """
+    @notice Recover Ether to recipient
+    """
+    amount: uint256 = self.balance
+    send(self.recipient, amount)
```

Comment:
Regular `send` might fail if ether will be transferred to a smart contract due to the 2300 gas limit.